### PR TITLE
builder: avoid parsing null- string to nil slice (issue #15634)

### LINKED ISSUE #15634: COPY fails when path starts with "null"?
I have a Dockerfile that contains the line

    COPY nullmailer /etc/nullmailer

and when I run `docker build .`, I get

    Step 15 : COPY 
    COPY requires at least two arguments

When I change the line to

    COPY hogenullmailer /etc/nullmailer

then it still fails (because there is no `hogenullmailer` file), but I get the correct error message:

    Step 15 : COPY hogenullmailer /etc/nullmailer
    hogenullmailer: no such file or directory

If I rename `nullmailer` to `hogenullmailer` then the process runs fine and I can build my image. Note that `COPY ["nullmailer", "/etc/nullmailer"]` also works.

    $ sudo docker version
    Client version: 1.7.1
    Client API version: 1.19
    Go version (client): go1.4.2
    Git commit (client): 786b29d
    OS/Arch (client): linux/amd64
    Server version: 1.7.1
    Server API version: 1.19
    Go version (server): go1.4.2
    Git commit (server): 786b29d
    OS/Arch (server): linux/amd64

This seems to be a recently introduced problem; I could build the same Dockerfile fine before the latest `apt-get dist-upgrade`.

### PR DESCRIPTION
**Pull request by skatsuta**
Fixes #15634

Currently parsing args in `ADD`, `COPY` and `VOLUME` instructions uses `Decoder.Decode()` first, and it binds `nil` if given bytes start with `null`.

This PR fixes to avoid parsing `null` string to `nil` slice by checking if the string starts with `[`.

Signed-off-by: Soshi Katsuta katsuta_soshi@cyberagent.co.jp

### CODE REVIEW

**Comment by estesp** [comment]
ping @duglin

**Comment by duglin** [comment]
@skatsuta overall it looks good.  Can you do me a  favor and add an integration-cli test too? I'd like to make sure that nothing else in the total flow does anything weird with args that start with "null".  It also ensures that if we rip out the parser and replace it with something else we'll still cover this case.

**Comment by skatsuta** [comment]
@duglin Thank you for your review!
I added a integration test. If there's another missing test context, please comment and I'll add it.

File: integration-cli/docker_cli_build_test.go [hunk 1]
```diff
@@ -5444,3 +5444,30 @@ func (s *DockerTrustSuite) TestBuildContextDirIsSymlink(c *check.C) {
...
+		COPY nullfile /
+		VOLUME nullvolume
+		`,
+		map[string]string{
+			"null":     "test1",
+			"nullfile": "test2",
+		},
+	)
+
+	if err != nil {
```
Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: You can replace this if-statement with: `c.Assert(err, check.IsNil)`  to be consistent with the other test clean-up PRs.

**Comment by duglin** [comment]
@skatsuta thanks for adding the test.
One minor comment, if you want to tweak it.
Either way, LGTM if janky is happy.

**Comment by skatsuta** [comment]
@duglin That statement is easier to read. I modified it to use `c.Assert()`.

**Comment by calavera** [comment]
LGTM

**Comment by duglin** [comment]
poor windows, needs some vitamin C or something  :-)

**Pull request merged**